frame/support/procedural/src/pallet/parse/extra_constants.rs

```diff
@@ -0,0 +1,121 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::helper;
+use syn::spanned::Spanned;
+
+/// List of additional token to be used for parsing.
+mod keyword {
+	syn::custom_keyword!(DispatchResultWithPostInfo);
+	syn::custom_keyword!(Call);
+	syn::custom_keyword!(OriginFor);
+	syn::custom_keyword!(weight);
+	syn::custom_keyword!(compact);
+	syn::custom_keyword!(T);
+	syn::custom_keyword!(pallet);
+}
+
+/// Definition of extra constants typically `impl<T: Config> Pallet<T> { ... }`
+pub struct ExtraConstantsDef {
+	/// The where_clause used.
+	pub where_clause: Option<syn::WhereClause>,
+	/// A set of usage of instance, must be check for consistency with trait.
+	pub instances: Vec<helper::InstanceUsage>,
+	/// The index of call item in pallet module.
+	pub index: usize,
+	/// The extra constant defined.
+	pub extra_constants: Vec<ExtraConstantDef>,
+}
+
+/// Input definition for an constant in pallet.
+pub struct ExtraConstantDef {
+	/// Name of the function
+	pub ident: syn::Ident,
+	/// The type returned by the function
+	pub type_: syn::Type,
+	/// The doc associated
+	pub doc: Vec<syn::Lit>,
+}
+
+impl ExtraConstantsDef {
```

Comment:
I want to question the reason for the syntax choice here. If I understood correctly, to define an extra constant in metadata you have to define a set of functions in an impl block. This sounds a bit misplaces, why not something like 

```
#[pallet_extra_const]
$vis const $ident: $type = $value
```

Reply:
Or maybe parsing `const fn` might be also a bit better.

Reply:
I think using const like

> ```rust
> #[pallet_extra_const]
> $vis const $ident: $type = $value
> ```

doesn't work because we need a way to declare the generics, and const doesn't support generics.

though we can do:

```rust
#[pallet_extra_const]
impl<T: Config> Pallet<T> {
  const FOO: T::Config = { $expr };
}
```

But then we have to delete or modify user input because a lot of stuff are not available in const environment (at least for now).

Thus I preferred user to declare in a way that the macro doesn't need to modify, but just needs to parse and use.

Though if it is not convincing I'm open to alternative

Reply:
Your reasoning is sound to me, I didn't think of generics tbh.